### DISCUSSION TITLE
Upgrading from 1.35.0 to 1.36.0 causes visibility values to not set properly

### DISCUSSION THREAD

**Discussion by alexpeverett** [bcakmakoglu/vue-flow#1513] [Q&A] [answered]
Hi,

We are using vue-flow for rendering graphs of workflows, their components, and the edges that connect them. Currently we have an issue where setting our hidden values for a node no longer has an effect. The issue was noticed when upgrading from 1.35.0 to 1.36.0 so we believe it may be an issue with the changes made in this update. There also appears to be an issue with the connectors not lining up with the nodes when we force the visibility of the node to true (shown below)

# Screenshots

## 1.35.0
<img width="907" alt="Screenshot 2024-07-01 at 10 16 39" src="https://github.com/bcakmakoglu/vue-flow/assets/43755376/d8072911-7b15-4088-a2f9-7ae8085a8494">
<img width="879" alt="Screenshot 2024-07-01 at 10 16 46" src="https://github.com/bcakmakoglu/vue-flow/assets/43755376/cb7fe6e6-f3b6-4cb7-a961-71310991054b">

## 1.36.0
<img width="905" alt="Screenshot 2024-07-01 at 10 13 32" src="https://github.com/bcakmakoglu/vue-flow/assets/43755376/26a2fb36-c41f-4189-9cc0-0e6ee8ab338f">
<img width="883" alt="Screenshot 2024-07-01 at 10 13 38" src="https://github.com/bcakmakoglu/vue-flow/assets/43755376/a544135a-216a-4851-bd2c-86d776e05fc3">

Force visibility on presets node:
![image](https://github.com/bcakmakoglu/vue-flow/assets/43755376/3506f399-5f7b-4265-b5e6-5d83d41fa6b8)

# HTML comparison (left new, right old)
![image (1)](https://github.com/bcakmakoglu/vue-flow/assets/43755376/b6e52901-4492-49ba-a35f-8a1cd67c4c1b)

**Comment by AlessandroPomponio** [comment]
Note: the HTML comparison was done with the following versions:
- New: 1.38.1
- Old: 1.33.8

**Reply by bcakmakoglu** [reply]
Are you dynamically showing/hiding handles? 🤔

**Reply by bcakmakoglu** [reply]
Also I can't reproduce the "hidden" issue. [The example works completely fine too.](https://vueflow.dev/examples/hidden.html)

**Reply by bcakmakoglu** [reply]
What changed from `1.35.0` to `1.36.0` is mainly that handles *will not auto generate an id anymore*.
So if you were relying on the fact that handles did create their own id automatically, you will have to resolve this by either removing those auto-generated ids from your already defined edges *or* set handle ids yourself.
The reasoning here is that auto-generated handle ids caused a lot of issues themselves so the pro's in terms of fixing bugs outweighed the cons of this sort of being a breaking change. It sucks, I know and I'd rather like to avoid that sort of thing in the future but for now it is what it is 😅

**Reply by AlessandroPomponio** [reply]
> Are you dynamically showing/hiding handles?

We are showing/hiding nodes and edges dynamically, see: https://registry.st4sd.res.ibm.com/registry-ui/experiment/nanopore-geometry-experiment/viewcanvas when clicking on "Show all inputs".

The way it's done is by doing the following: https://github.com/st4sd/st4sd-registry-ui/blob/7eef649ed59652fdb3c68a2a2dc376cdfcc758e8/src/canvas/components/canvases/ViewCanvas.vue#L187-L201

```js
let Presets = nodes.value.find((a) => a.id == "presets");
  if (Presets != undefined) {
    Presets.hidden = !inputNodesVisability.value;
  }
```

By logging to console, I can see that the `find` method finds the node and the value is updated accordingly so I'm not sure the IDs should be the issues at play

**Reply by AlessandroPomponio** [reply]
This is also the definition of the Presets input node https://github.com/st4sd/st4sd-registry-ui/blob/7eef649ed59652fdb3c68a2a2dc376cdfcc758e8/src/canvas/functions/canvasFunctions.js#L63-L77: 

```js
function addPresetsInputNode(graph) {
  return graph.nodes.push({
    id: "presets",
    label: "Presets",
    definition: {},
    parentNode: "",
    type: "input",
    hidden: true,
    style: {
      borderColor: "red",
    },
    position: { x: 200, y: 0 },
    width: getTextWidth("presets"),
  });
}
```

**Reply by bcakmakoglu** [reply]
I need a proper (minimal) reproduction of the issue since I can't reproduce it myself and the code snippets aren't telling me that much + I can't spend the time exploring the repo itself :/

[You can find a sandbox template right here.](https://codesandbox.io/p/sandbox/vue-flow-basic-gfgro4)

**Reply by bcakmakoglu** [reply]
Slightly off-topic but you're going through, imo, unnecessary hoops to achieve what you're trying to do in this code-block

```ts
const showHideInputNodes = () => {
  inputNodesVisability.value = !inputNodesVisability.value;
  let Other = nodes.value.find((a) => a.id == "other");
  if (Other != undefined) {
    Other.hidden = !inputNodesVisability.value;
  }
  let Presets = nodes.value.find((a) => a.id == "presets");
  if (Presets != undefined) {
    Presets.hidden = !inputNodesVisability.value;
  }
  showHideButtonTitle.value =
    showHideButtonTitle.value == "Hide presets & other"
      ? "Show all inputs"
      : "Hide presets & other";
};
```

You can clean this up to something like this and avoid having to loop through n-nodes to do an update on a specific one.

```ts
const { updateNode } = useVueFlow()

function showHideInputNodes() {
  const isVisible = inputNodesVisibility.value

	// loop through the different node ids to update
  ;['other', 'presets'].forEach((id) => updateNode(id, { hidden: isVisible }))

	// alternatively use this
	updateNode('other', { hidden: isVisible })
	updateNode('presets', { hidden: isVisible })

  showHideButtonTitle.value = showHideButtonTitle.value === 'Hide presets & other' ? 'Show all inputs' : 'Hide presets & other'
  inputNodesVisibility.value = !inputNodesVisibility.value
}
```

There are a number of built-in helpers that would make your life much easier like `findNode(nodeId)` which avoids having to loop through a whole array to get a specific node (using a map internally).

Also helpers like `updateNode` that can already handle the finding and updating of a nodes properties for you, again without having to loop through a whole array to do so :)

**Reply by AlessandroPomponio** [reply]
Thank you @bcakmakoglu, this is super helpful! Unfortunately this code was written when some of the people were still getting familiar with your (lovely) framework and Vue/JS in general, so lots of things should be refactored, but time is always way too limited 😢 

Hopefully we'll find some time for it soon!

In the meantime, here's a reproducible example for you: https://codesandbox.io/p/devbox/infallible-bose-pvm574
After writing the function, everything seemed to be working and I was already scratching my head, but I noticed that the `@vue-flow/core` package was pinned to an old version 👍

**Reply by bcakmakoglu** [reply]
Thanks for the repro, the bug will be fixed in the next patch 👍

**Reply by bcakmakoglu** [reply, marked as answer]
Fixed with `1.38.2`

**Reply by bcakmakoglu** [reply]
btw you *are allowed* to open actual issues if you suspect that a bug is happening 😄
Just as an info. I like to keep Issues solely for bugs/problems while Discussions can include anything, so I won't stop you from opening discussions either ^^

**Reply by AlessandroPomponio** [reply]
Thank you very much @bcakmakoglu !
